src/hotspot/share/opto/opaquenode.cpp

```diff
@@ -107,27 +107,59 @@ const Type* OpaqueNotNullNode::Value(PhaseGVN* phase) const {
 }
 
 Node* OpaqueTemplateAssertionPredicateNode::Identity(PhaseGVN* phase) {
-  if (phase->C->post_loop_opts_phase()) {
-    // Template Assertion Predicates only serve as templates to create Initialized Assertion Predicates when splitting
-    // a loop during loop opts. They are not used anymore once loop opts are over and can then be removed. They feed
-    // into the bool input of an If node and can thus be replaced by true to let the Template Assertion Predicate be
-    // folded away (the success path is always the true path by design).
-    return phase->intcon(1);
-  } else {
+  if (!phase->C->post_loop_opts_phase()) {
+    // Record Template Assertion Predicates for post loop opts IGVN. We can remove them when there is no more loop
+    // splitting possible. This also means that we do not create any new Initialized Assertion Predicates created from
+    // these templates.
     phase->C->record_for_post_loop_opts_igvn(this);
   }
   return this;
 }
 
 const Type* OpaqueTemplateAssertionPredicateNode::Value(PhaseGVN* phase) const {
+  if (_useless || phase->C->post_loop_opts_phase()) {
+    // Template Assertion Predicates only serve as templates to create Initialized Assertion Predicates when splitting
+    // a loop during loop opts. They are not used anymore once loop opts are over and can then be removed. They feed
+    // into the bool input of an If node and can thus be replaced by the success path to let the Template Assertion
+    // Predicate be folded away (the success path is always the true path by design). We can also fold the Template
+    // Assertion Predicate away when it's found to be useless and not used anymore.
+    return TypeInt::ONE;
+  }
   return phase->type(in(1));
 }
 
+#ifndef PRODUCT
+void OpaqueTemplateAssertionPredicateNode::dump_spec(outputStream* st) const {
+  if (_useless) {
+    st->print("#useless ");
+  }
+}
+#endif // NOT PRODUCT
+
+// Avoid that a useless OpaqueInitializedAssertionPredicateNode is commoned up with a useful one.
+bool OpaqueInitializedAssertionPredicateNode::cmp(const Node &n) const {
+  return _useless == n.as_OpaqueInitializedAssertionPredicate()->is_useless();
+}
+
 const Type* OpaqueInitializedAssertionPredicateNode::Value(PhaseGVN* phase) const {
+  if (_useless) {
+    return TypeInt::ONE;
+  }
   return phase->type(in(1));
 }
 
-//=============================================================================
+void OpaqueInitializedAssertionPredicateNode::mark_useless(PhaseIterGVN& igvn) {
+  _useless = true;
+  igvn._worklist.push(this);
+}
+
+#ifndef PRODUCT
+void OpaqueInitializedAssertionPredicateNode::dump_spec(outputStream* st) const {
```

Comment:
Added dump for better debugging, also done for `OpaqueTemplateAssertionPredicateNode`.